### ISSUE TITLE
Add StartTLS support to LDAP

### ISSUE THREAD

**Issue by miketweaver** [thelounge/thelounge#2403]
Currently, thelounge does not support StartTLS for the ldap:// protocol. 

It's on the `TODO` comments for `ldap.js`:
https://github.com/thelounge/thelounge/blob/16236ec0bf04f6a1f375e54d69e64124e4f89149/src/plugins/auth/ldap.js#L111

It was also talked about in the PR https://github.com/thelounge/thelounge/pull/980#issuecomment-318982203 from @eliemichel, which added the LDAP support but it was never added.

`ldapjs` does support [starttls()](http://ldapjs.org/client.html#starttls) but the feature will need to be integrated into this project.

I would like to suggest that we add StartTLS support, as ldaps:// is deprecated in favor of StartTLS [[RFC2830](http://www.rfc-base.org/txt/rfc-2830.txt)].

**Comment by astorije** [comment]
/Cc @gramakri, in case you have any free time (does that even exist? 😁)

**Comment by lol768** [comment]
> I would like to suggest that we add StartTLS support, as ldaps:// is deprecated in favor of StartTLS [RFC2830].

How on earth did this happen? Urgh, I hate StartTLS and everything it stands for - it was dumb in SMTP/IMAP, it was dumb in IRCv3 (and good client libraries [refuse to support it](https://github.com/KittehOrg/KittehIRCClientLib/blob/4cc34feac1c7da3a1cdae92a32eb891224e2378b/docs/ircv3.md)) and it's dumb here in LDAP.

It absolutely does more harm than good due to it becoming completly useless in the face of an active MitM (I've seen way too many broken clients happily carry on in plaintext if their STARTTLS command got ignored!), introducing further complexity and being _no more secure than a TLS port_. These protocols should just enforce TLS as a pre-requisite for all connections.

https://www.fastmail.com/help/technical/ssltlsstarttls.html?domain=fastmail.fm

https://www.agwa.name/blog/post/starttls_considered_harmful

**Comment by miketweaver** [comment]
@lol768 I agree with you 100%, but StartTLS has been a part of the spec for at least 12 years now (since LDAPv3) and should be supported over support for LDAPS, as LDAPS is not really specified anywhere, nor even mentioned, in the LDAP RFC. (_I know it's commonly used by various applications._) By supporting StartTLS with LDAPv3 (along side the current implementation of LDAPS), you would allow users to use a mechanism that is actually specified in the RFC. 

To be transparent here, my goals here are selfish. I am hoping that we support StartTLS because is the only encryption method supported by the LDAP server I'm trying to connect with thelounge. _On a side note, I wish something could be done about the issues you brought up, but I don't think it's this projects place or in its' goals to try to push changes to the LDAP protocol._

@xPaw 
Here's an example chat program using `starttls`:
https://github.com/RocketChat/Rocket.Chat/blob/582f5a0636f0e84960151d7bddbd9a672e19e157/packages/rocketchat-ldap/server/ldap.js#L134-L169

Other projects require explicit declarations of the encryption type.
Example ([Portainer.io](https://portainer.io/)):
![image](https://user-images.githubusercontent.com/5581855/39677585-f1d2f4da-513a-11e8-8803-0a88c8a6a949.png)

Both examples have the user explicitly select the encryption type.
Hopefully that answers your question?
